### ISSUE TITLE
Failed to start all services modules

### ISSUE THREAD

**Issue by VironicaZ** [bioinfo-chru-strasbourg/STARK#5]
Dear STARK team,

I get this error when i try to run "sudo services/services.sh --modules=* --command=up":

#######################################
 STARKServicesModules [0.9.1b-01/10/2020]
 STARK Services Modules 
 Antony Le Bechec @ HUS/CPS © GNU GPLA V3
#######################################
#[INFO] STARK Module Docker version 'Docker version 20.10.12, build 20.10.12-0ubuntu4'
#[INFO] STARK Module Docker Compose environment file '/home/vironica/STARK/services/../.env'
#[INFO] STARK Module Docker Compose environment file '/home/vironica/STARK/services/STARK.env'
#[INFO] STARK Module main folder '//root/STARK'
#[INFO] STARK Module services modules folder '//root/STARK/services'
#[INFO] STARK Module Docker Compose Command 'up -d'
#[INFO] STARK Module Common services 'common'

#[INFO] STARK Module 'STARK'
#[INFO] STARK Module 'STARK' SubModule 'main'
/snap/docker/2343/lib/python3.6/site-packages/paramiko/transport.py:33: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography and will be removed in a future release.
  from cryptography.hazmat.backends import default_backend
#[ERROR] docker-compose error2
/snap/docker/2343/lib/python3.6/site-packages/paramiko/transport.py:33: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography and will be removed in a future release.
  from cryptography.hazmat.backends import default_backend
Couldn't find env file: /tmp/STARK/services.18221/.env

Moreover, i used the more complete install procedure.

**My OS:** 
Ubuntu 22.04.1 LTS

**My docker version:** 
20.10.12

Thanks in advance for your help.

Best,

**Comment by antonylebechec** [comment]
Dear Vironica,

This error may be due to the deprecated "docker-compose" command, used in "services.sh" script. 

I suggest to create an alias to new docker compose command:
```
alias docker-compose='docker compose'
```
as mentioned in the following thread:
https://askubuntu.com/questions/1439553/how-to-silence-or-fix-snap-docker-compose-cryptographydeprecationwarning-python

Future release of STARK will take into account this new command of docker.

Best regards,

**Comment by Nour-EddineS** [comment]
Hello @antonylebechec,

I don't have "docker compose" on my docker version, that's what I have "docker-compose"!
So if I understand correctly I have to change the docker version which contains "docker compose" and not "docker-compose"?
The version i have:

$ docker-compose version

/snap/docker/2343/lib/python3.6/site-packages/paramiko/transport.py:33: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography and will be removed in a future release.
from cryptography.hazmat.backends import default_backend
docker-compose version 1.29.2, build unknown
docker-py version: 5.0.3
CPython version: 3.6.9
OpenSSL version: OpenSSL 1.1.1  11 Sep 2018

Sincerely,

**Comment by VironicaZ** [comment]
Dear @antonylebechec,

I think it's not my case, because **"docker compose"** is in the new version of docker, for me I have the version that contains **"docker-compose"**.
Should I migrate to the new version of docker?

Best regards,

**Comment by antonylebechec** [comment]
Dear Vironica,

STARK is supposed to work with this version of Docker. So I think you don't have to update your version of Docker.

Moreover, the error message "Couldn't find env file: /tmp/STARK/services.18221/.env" come from docker-compose command. It means that the command is launch, your docker-compose seems ok.

It could be a problem due to file access. Modules are configured with a ".env" file, with an inheritance process. That's why a temporary ".env" file is generated, from the merge of multiple configuration files. 

Can you check your folders? Especially:
/home/vironica/STARK
/root/STARK

Depending on where you setup STARK scripts, you can specify services folder (e.g. /home/vironica/STARK/services):
```
services.sh --modules=* --command=up --folder_services=/home/vironica/STARK/services --debug
```

Can you launch your commands in debug mode (--debug)?

We plan to develop a new release of STARK. All bugs and suggestions are very useful! So, thank you for your feedback!

Best,

**Comment by Nour-EddineS** [comment]
Dear @antonylebechec,

Always the same problem.

I launched the command in debug mode; 

`sudo services/services.sh --modules=* --command=up --folder_services=/home/stark/STARK/services --debug`

see the command output in console.log file.

[console.log](https://github.com/bioinfo-chru-strasbourg/STARK/files/10455983/console.log)

Best regards,

**Comment by antonylebechec** [comment]
Dear @Nour-EddineS,

Can you provide the tmp folder "/tmp/STARK/services.9551"? (as a zip file)

I strongly suspect a tmp file access configuration ("Couldn't find env file: /tmp/STARK/services.9551/.env"). This ".env" file can be read, but probably can not be executed (which is need for docker-compose). Can you provide the output of this following command?
```
ls -l /tmp/STARK/services.9551/*
```

In order to check, can you provide the output of these following commands?
```
chmod 0777 -R /tmp/STARK/services.9551/*
chmod 0777 -R /home/stark/STARK/services/STARK/*
docker-compose --file /home/stark/STARK/services/STARK/STARK.docker-compose.yml --env-file /tmp/STARK/services.9551/.env -p STARK
```

Thanks for your help!

Best,

**Comment by Nour-EddineS** [comment]
Dear @antonylebechec,

Thank you very much for your responsiveness.
I share with you the tmp folder as a zip file("tmp_folder.zip").
[tmp_folder.zip](https://github.com/bioinfo-chru-strasbourg/STARK/files/10458374/tmp_folder.zip)

The result of the command:
`ls -l /tmp/STARK/*`
/tmp/STARK/services.15947:
total 4
-rw-r--r-- 1 root root 374 Jan 19 10:50 docker-compose.err
-rw-r--r-- 1 root root   0 Jan 19 10:50 docker-compose.log

/tmp/STARK/services.1898:
total 4
-rw-r--r-- 1 root root 373 Jan 19 10:54 docker-compose.err
-rw-r--r-- 1 root root   0 Jan 19 10:54 docker-compose.log

/tmp/STARK/services.22571:
total 4
-rw-r--r-- 1 root root 374 Jan 19 10:56 docker-compose.err
-rw-r--r-- 1 root root   0 Jan 19 10:56 docker-compose.log

/tmp/STARK/services.22884:
total 4
-rw-r--r-- 1 root root 374 Jan 19 10:50 docker-compose.err
-rw-r--r-- 1 root root   0 Jan 19 10:49 docker-compose.log

/tmp/STARK/services.29946:
total 4
-rw-r--r-- 1 root root 374 Jan 19 11:09 docker-compose.err
-rw-r--r-- 1 root root   0 Jan 19 11:09 docker-compose.log

/tmp/STARK/services.9551:

total 4
-rw-r--r-- 1 root root 373 Jan 19 11:03 docker-compose.err
-rw-r--r-- 1 root root   0 Jan 19 11:03 docker-compose.log

The result of the command:
`ls -l /tmp/STARK/services.9551/*`

total 4
-rw-r--r-- 1 root root 373 Jan 19 11:03 docker-compose.err
-rw-r--r-- 1 root root   0 Jan 19 11:03 docker-compose.log

The result of the command:
`docker-compose --file /home/stark/STARK/services/STARK/STARK.docker-compose.yml --env-file /tmp/STARK/services.9551/.env -p STARK config`
/snap/docker/2343/lib/python3.6/site-packages/paramiko/transport.py:33: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography and will be removed in a future release.
from cryptography.hazmat.backends import default_backend
ERROR: Couldn't find env file: /tmp/STARK/services.9551/.env

Thank you in advance.

Best regards,

**Comment by VironicaZ** [comment]
Dear @antonylebechec,

Does the command `docker-compose --project-name STARK up stark-databases`
takes a long time to complete?

In my case almost 16 hours and not yet finished, it's normal!. Therefore will this affect the start of services?

Best,

**Comment by antonylebechec** [comment]
Dear @VironicaZ,

Yes, the database generation step takes a long time. It's a step that download multiple databases (genome, annotations files, dbSNP, refGene...), build indexes, format some files in standards.

By default, this step use 4 threads, which takes a while. A solution consists on modify ".env" to change the following parameter (depend on your infrastructure):
```
DOCKER_STARK_GET_DATABASES_THREADS=4
```

This step need to finish before using services, as databases are needed for the analyses. Moreover, services checked if databases are well download/configured to be considered as "healthy".

Fortunately, this step is done only once, at the setup of STARK.

Best,

**Comment by antonylebechec** [comment]
Dear @Nour-EddineS,

Thank you for your reply.

First of all, it seems that the Python warning can failed the command (not sure yet, I checked Docker docs but not found answer). You should probably update your Python to calm Docker!

Moreover, I'm currently not able to reproduce the "env file" error. So, I'll will ask you to do some more check...

Can you tell me which user you use? I guess root?

First, I need lists with hidden files ("ls -l**a**"):
```
ls -la /tmp/STARK/*
ls -la /tmp/STARK/services.9551/*
```

Then, change file access:
```
chmod 0777 -R /tmp/STARK
```

Finally, the command:
```
docker-compose --file /home/stark/STARK/services/STARK/STARK.docker-compose.yml --env-file /tmp/STARK/services.9551/.env -p STARK config
```

I suggest a temporary fix by using the services script provided by "STARK-modules" github project:
https://github.com/bioinfo-chru-strasbourg/STARK-modules
Just replace file "services/services.sh", and let me know if it works.

@VironicaZ, do you have the same error?

I'm sorry for the inconvenience.

Best,

**Comment by Nour-EddineS** [comment]
Dear @antonylebechec,

The result of **ls -la /tmp/STARK/*** :
`
/tmp/STARK/services.24268:
total 28
drwxr-xr-x 2 root root  4096 Jan 20 08:44 .
drwxr-xr-x 4 root root  4096 Jan 20 08:51 ..
-rw-r--r-- 1 root root   374 Jan 20 08:44 docker-compose.err
-rw-r--r-- 1 root root     0 Jan 20 08:44 docker-compose.log
-rw-r--r-- 1 root root 13265 Jan 20 08:44 .env
-rw-r--r-- 1 root root     0 Jan 20 08:44 .env.err

/tmp/STARK/services.25099:
total 28
drwxr-xr-x 2 root root  4096 Jan 20 08:51 .
drwxr-xr-x 4 root root  4096 Jan 20 08:51 ..
-rw-r--r-- 1 root root   374 Jan 20 08:51 docker-compose.err
-rw-r--r-- 1 root root     0 Jan 20 08:51 docker-compose.log
-rw-r--r-- 1 root root 13265 Jan 20 08:51 .env
-rw-r--r-- 1 root root     0 Jan 20 08:51 .env.err

/tmp/STARK/services.11142:
total 28
drwxr-xr-x 2 root root  4096 Jan 20 11:11 .
drwxr-xr-x 4 root root  4096 Jan 20 11:11 ..
-rw-r--r-- 1 root root   374 Jan 20 11:11 docker-compose.err
-rw-r--r-- 1 root root     0 Jan 20 11:11 docker-compose.log
-rw-r--r-- 1 root root 13265 Jan 20 11:11 .env
-rw-r--r-- 1 root root     0 Jan 20 11:11 .env.err

/tmp/STARK/services.9551:
total 28
drwxr-xr-x 2 root root  4096 Jan 19 11:03  .
drwxr-xr-x 4 root root  4096 Jan 19 11:03  ..
-rw-r--r-- 1 root root   373 Jan 19 11:03  docker-compose.err
-rw-r--r-- 1 root root     0 Jan 19 11:03  docker-compose.log
-rw-r--r-- 1 root root 13265 Jan 19 11:03  .env
-rw-r--r-- 1 root root     0 Jan 19 11:03 .env.err
`
The result of **ls -la /tmp/STARK/services.9551/*** :
`-rw-r--r-- 1 root root 374 Jan 19 11:03 /tmp/STARK/services.9551/docker-compose.err
-rw-r--r-- 1 root root   0 Jan 19 11:03 /tmp/STARK/services.9551/docker-compose.log`

The result of the command (after the change change the access permissions "chmod 0777 -R /tmp/STARK"):
`docker-compose --file /home/stark/STARK/services/STARK/STARK.docker-compose.yml --env-file /tmp/STARK/services.9551/.env -p STARK config`

/snap/docker/2343/lib/python3.6/site-packages/paramiko/transport.py:33: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography and will be removed in a future release.
from cryptography.hazmat.backends import default_backend
ERROR: Couldn't find env file: /tmp/STARK/services.9551/.env

I use normal user but i run command with sudo.

Also I used your proposal(the services script provided by "STARK-modules"), but the same problem :( !

Best regards,

**Comment by antonylebechec** [comment]
Dear @Nour-EddineS,

This is strange. The ".env" file can be read, but docker-compose can't not read it. Following Docker help, the command is correct:
https://docs.docker.com/compose/environment-variables/
Moreover, it works in multiple environments we tested (MacOS, Linux/CentOS, Linux/Ubuntu...). However, we didn't tested all environments and Docker/Python releases...

Can you copy both files in a folder?
```
cp /home/stark/STARK/services/STARK/STARK.docker-compose.yml <folder>/docker-compose.yml
cp /tmp/STARK/services.9551/.env <folder>/.env
cd <folder>
chmod 0777 *
```
and run these different following commands?
```
docker-compose config
docker-compose --file docker-compose.yml --env-file .env config
docker-compose --file ./docker-compose.yml --env-file ./.env config
```

Do you already use this kind of command? I mean docker-compose with --env-file option, in another project?

Another way to deal with env variables with docker compose is to use this command:
```
env $(cat /tmp/STARK/services.9551/.env | grep "#" -v) docker-compose --file /home/stark/STARK/services/STARK/STARK.docker-compose.yml -p STARK config
```
Basically, "services.sh" check if the option "--env-file" exists and use it, or use this alternative command if not exists. It's a fix for old releases of docker compose. 

Let me know the output of these commands.

Best,

**Comment by Nour-EddineS** [comment]
Dear @antonylebechec,

 Output of :
`docker-compose config
`
[consol1.log](https://github.com/bioinfo-chru-strasbourg/STARK/files/10467400/consol1.log)

Output of :
`docker-compose --file docker-compose.yml --env-file .env config`

/snap/docker/2343/lib/python3.6/site-packages/paramiko/transport.py:33: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography and will be removed in a future release.
from cryptography.hazmat.backends import default_backend
services: {}
version: '3'

Output of :
`env $(cat /tmp/STARK/services.9551/.env | grep "#" -v) docker-compose --file /home/stark/STARK/services/STARK/STARK.docker-compose.yml -p STARK config`

/snap/docker/2343/lib/python3.6/site-packages/paramiko/transport.py:33: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography and will be removed in a future release.
from cryptography.hazmat.backends import default_backend
services: {}
version: '3'

Sincerely,

**Comment by VironicaZ** [comment]
Dear @antonylebechec,

Yes i have same error. Also, I have implemented your comments that you requested for @Nour-EddineS  . However, the same problem !!!

**Comment by antonylebechec** [comment]
Dear @Nour-EddineS, @VironicaZ,

I suggest a quick and temporary fix.

In "services/services.sh" file, modify line 467:
```
if (( $(docker-compose --help | grep "\-\-env\-file" -c) )); then
```
by
```
if (( $(docker-compose --help | grep "\-\-env\-file" -c) )) && ((0)); then
```

This fix will force alternative way to use env variables.

Let me know!

Best,

**Comment by Nour-EddineS** [comment]
Dear @antonylebechec,
I run : 
`sudo services/services.sh --modules=* --command=up --debug`
Output of command : 
[output.log](https://github.com/bioinfo-chru-strasbourg/STARK/files/10468030/output.log)

Best,

**Comment by VironicaZ** [comment]
Dear @antonylebechec,

I run the command ( **services/services.sh --modules=* --command=up** ), his is what I get as a message:

#[WARNING] STARK Module Docker version does not provide '--env-file' parameter. STARK module is patched to work, but please update your docker version.
#[ERROR] docker-compose error
env: ‘"while’: No such file or directory

Best,

**Comment by antonylebechec** [comment]
Dear @Nour-EddineS, @VironicaZ,

The alternative way work only with simple env file. Unfortunately, the env file is more complex than expected (variables with command that are interpreted). 

The different tests executed show that it should be a problem with docker compose, which not correctly deal with --env-file option in your system. This may be due to Docker compose version or Python version (see the error/warning message).

Here are some threads that can help you:
https://www.bellavrforum.org/t/jetson-nano-docker-problems/281
https://serverfault.com/questions/1033128/cryptographydeprecationwarning-python-2-is-no-longer-supported-by-the-python-co

Currently, I'm not able to fix this issue, as it seems to not be specifically a STARK issue.

Best,

**Comment by VironicaZ** [comment]
Dear @antonylebechec,
You don't have a virtual linux machine contains STARK installed ?

Kind regards,

**Comment by antonylebechec** [comment]
Dear @VironicaZ,

We do not provide a virtual machine. This is mainly due to the size of the databases. That's why we switched to Docker, which is a useful virtualisation solution.

However, it's a good solution to install STARK inside a virtual machine, for example with Ubuntu 20.04 LTS.

Best,

**Comment by Nour-EddineS** [comment]
Dear @antonylebechec,

I tried installing STARK on a virtual machine, Ubuntu 22.04 LTS. With the latest version of docker which contains "**docker compose**" instead of "**docker-compose**".
So I started the **build** step with: "docker compose build"; step completed successfully.
For the **setup** step here are the commands that I launched:

1) `DOCKER_STARK_MAIN_FOLDER=${HOME}/STARK`
2)` mkdir -p $DOCKER_STARK_MAIN_FOLDER`

When I tried to run these three commands as written below, I got an error message that said ""STARK" is not a valid project name".

3) `docker compose --project-name STARK up stark-setup`
4)` docker compose --project-name STARK up stark-databases`
5) `docker compose --project-name STARK up stark-sources-archives`

To solve this problem I eliminated "--project-name STARK " (I don't know if it's correct). Therefore the commands completed successfully.
The commands that I launched after modification:
3') `docker compose up stark-setup`
4') `docker compose up stark-databases`
5') `docker compose up stark-sources-archives.`

So to launch the Services step I changed in the services/services.sh file the "docker-compose" commands with the new "docker compose".
When I ran the command:
`services/services.sh --modules=* --command=up --debug`

I got an error message, see the output.log file.
[output.log](https://github.com/bioinfo-chru-strasbourg/STARK/files/10478813/output.log)

Best regards,

**Comment by antonylebechec** [comment]
Dear @Nour-EddineS,

Due to new spec of docker compose, option --project-name is now case sensitive. This option is needed to create STARK network with the correct name. So, you can run these commands (only if network "stark_stark" is not created, but if you run these commands in a "stark" folder, the project will be "stark" by default):

```
docker compose --project-name stark up stark-setup
docker compose --project-name stark up stark-databases
docker compose --project-name stark up stark-sources-archives
```

To fix the STARK services script, 2 solutions:
- change case of each service folder:
```
services/STARK
```
to
```
services/stark
```
- change code in services.sh (line 327):
```
module_name=$(basename $service_module)
```
to
```
module_name=$(basename $service_module | awk '{print tolower($0)}')
```

In a future release, these 2 fixes will be distributed.

**Comment by Nour-EddineS** [comment]
Dear @antonylebechec,

Finally after this last remark the installation completed successfully.
Hereby, I would like to thank you for your help.
Your input was so valuable to me. Thanks to your help, the problem is solved. Thank you for giving me your time to help me.

So I can start the NGS data analysis stage!

With all my gratitude.
Sincerely,

**Comment by VironicaZ** [comment]
Dear @antonylebechec,

Great, services is starting :) 
Thank you for your support. Your wisdom is always a great way to control error.

Best regards,

**Comment by antonylebechec** [comment]
Dear @Nour-EddineS, @VironicaZ,

Thank you for your input!

This threads help us to identify some future fixes and improvements (we'll open issues to follow these issues). Especially, an issue on module name case, and another on how to install STARK on a virtual machine. Would it be possible for you both to share you experience on virtual machine?

Your next step is to launch an analysis with STARK. Let us know if you need help!

Moreover, I suggest you to have a look on additional for STARK:
https://github.com/bioinfo-chru-strasbourg/STARK-modules

Best regards,
